File: src/variant_base.cpp
#include <protean/variant.hpp>
#include <protean/detail/hash.hpp>
#include <protean/detail/data_table.hpp>

#include <boost/algorithm/string/predicate.hpp>

namespace protean {

    variant_base::variant_base()
    {
    }

    void variant_base::initialise(enum_type_t type, size_t size)
    {
        if ((type & Number)!=0)
        {
            m_value.set<UInt64>(0);
        }
        else
        {
            switch( type )
            {
                case None:
                    m_value.set<None>(boost::blank());
                    break;
                case Date:
                    m_value.set<Date>(boost::gregorian::date());
                    break;
                case Time:
                    m_value.set<Time>(boost::posix_time::time_duration());
                    break;
                case DateTime:
                    m_value.set<DateTime>(boost::posix_time::ptime());
                    break;
                case TimeSeries:
                    m_value.set<TimeSeries>(detail::timeseries());
                    break;
                case DataTable:
                    m_value.set<DataTable>(detail::data_table(size));
                    break;
                case Any:
                    m_value.set<Any>(detail::string());
                    break;
                case String:
                    m_value.set<String>(detail::string());
                    break;
                case Dictionary:
                    m_value.set<Dictionary>(detail::dictionary());
                    break;
                case Bag:
                    m_value.set<Bag>(detail::bag());
                    break;
                case List:
                    m_value.set<List>(detail::list(size));
                    break;
                case Tuple:
                    m_value.set<Tuple>(detail::tuple(size));
                    break;
                case Buffer:
                    m_value.set<Buffer>(handle<detail::buffer>(new detail::buffer(size)));
                    break;
                default:
                    boost::throw_exception(variant_error(std::string("Cannot default construct variant of type ") + enum_to_string(type)));
            }
        }
    }

    variant_base::variant_base(enum_type_t type, const variant_base& rhs)
    {
        assign(type, rhs);
    }

    void variant_base::assign(enum_type_t type, const variant_base& rhs)
    {
        if ((type & Number)!=0)
        {
            m_value.set<UInt64>(rhs.m_value.get<UInt64>());
        }
        else
        {
            switch(type)
            {
                case None:
                    m_value.set<None>(rhs.m_value.get<None>());
                    break;
                case Any:
                    m_value.set<Any>(rhs.m_value.get<Any>());
                    break;
                case String:
                    m_value.set<String>(rhs.m_value.get<String>());
                    break;
                case Date:
                    m_value.set<Date>(rhs.m_value.get<Date>());
                    break;
                case Time:
                    m_value.set<Time>(rhs.m_value.get<Time>());
                    break;
                case DateTime:
                    m_value.set<DateTime>(rhs.m_value.get<DateTime>());
                    break;
                case List:
                    m_value.set<List>(rhs.m_value.get<List>());
                    break;
                case Dictionary:
                    m_value.set<Dictionary>(rhs.m_value.get<Dictionary>());
                    break;
                case Bag:
                    m_value.set<Bag>(rhs.m_value.get<Bag>());
                    break;
                case TimeSeries:
                    m_value.set<TimeSeries>(rhs.m_value.get<TimeSeries>());
                    break;
                case DataTable:
                    m_value.set<DataTable>(rhs.m_value.get<DataTable>());
                    break;
                case Buffer:
                    m_value.set<Buffer>(rhs.m_value.get<Buffer>());
                    break;
                case Tuple:
                    m_value.set<Tuple>(rhs.m_value.get<Tuple>());
                    break;
                case Object:
                    m_value.set<Object>(rhs.m_value.get<Object>());
                    break;
                case Exception:
                    m_value.set<Exception>(rhs.m_value.get<Exception>());
                    break;
                case Array:
                    m_value.set<Array>(rhs.m_value.get<Array>());
                    break;
                default:
                    boost::throw_exception(variant_error(std::string("Attempt to copy unknown variant type ") + enum_to_string(type)));
            }
        }
    }

    void variant_base::destroy(enum_type_t type)
    {
        if ((type & Number)!=0)
        {
            m_value.set<UInt64>(0);
        }
        else
        {
            switch( type )
            {
                case None:
                    m_value.destroy<None>();
                    break;
                case Any:
                    m_value.destroy<Any>();
                    break;
                case String:
                    m_value.destroy<String>();
                    break;
                case Date:
                    m_value.destroy<Date>();
                    break;
                case Time:
                    m_value.destroy<Time>();
                    break;
                case DateTime:
                    m_value.destroy<DateTime>();
                    break;
                case List:
                    m_value.destroy<List>();
                    break;
                case Dictionary:
                    m_value.destroy<Dictionary>();
                    break;
                case Bag:
                    m_value.destroy<Bag>();
                    break;
                case TimeSeries:
                    m_value.destroy<TimeSeries>();
                    break;
                case DataTable:
                    m_value.destroy<DataTable>();
                    break;
                case Buffer:
                    m_value.destroy<Buffer>();
                    break;
                case Tuple:
                    m_value.destroy<Tuple>();
                    break;
                case Object:
                    m_value.destroy<Object>();
                    break;
                case Exception:
                    m_value.destroy<Exception>();
                    break;
                case Array:
                    m_value.destroy<Array>();
                    break;
                default:
                    boost::throw_exception(variant_error(std::string("Attempt to destruct unknown variant type ") + enum_to_string(type)));
            }
        }
    }

    std::string variant_base::enum_to_string(enum_type_t type)
    {
        return variant_impl_t::enum_to_string(type);
    }
    variant_base::enum_type_t variant_base::string_to_enum(const std::string& type)
    {
        return static_cast<variant_base::enum_type_t>(variant_impl_t::string_to_enum(type));
    }

    template<typename T>
    static int compare_using_less(const T& lhs, const T& rhs)
    {
        if (lhs==rhs)
        {
            return 0;
        }
        else
        {
            return lhs<rhs ? -1 : 1;
        }
    }

    int variant_base::compare(enum_type_t type, const variant_base& rhs) const
    {
        // structures are of same type
        switch (type)
        {
            case None:
                return 0;
            case Any:
                return m_value.get<Any>().compare(rhs.m_value.get<Any>());
            case String:
                return m_value.get<String>().compare(rhs.m_value.get<String>());
            case Int32:
                return compare_using_less(m_value.get<Int32>(), rhs.m_value.get<Int32>());
            case UInt32:
                return compare_using_less(m_value.get<UInt32>(), rhs.m_value.get<UInt32>());
            case Int64:
                return compare_using_less(m_value.get<Int64>(), rhs.m_value.get<Int64>());
            case UInt64:
                return compare_using_less(m_value.get<UInt64>(), rhs.m_value.get<UInt64>());
            case Float:
                return compare_using_less(m_value.get<Float>(), rhs.m_value.get<Float>());
            case Double:
                return compare_using_less(m_value.get<Double>(), rhs.m_value.get<Double>());
            case Boolean:
                if ( m_value.get<Boolean>()==rhs.m_value.get<Boolean>() )
                    return 0;
                else if ( m_value.get<Boolean>() )
                    return 1;
                else
                    return -1;
                break;
            case Date:
                return compare_using_less(m_value.get<Date>(), rhs.m_value.get<Date>());
            case Time:
                return compare_using_less(m_value.get<Time>(), rhs.m_value.get<Time>());
            case DateTime:
                return compare_using_less(m_value.get<DateTime>(), rhs.m_value.get<DateTime>());
            case Buffer:
                return m_value.get<Buffer>()->compare(*rhs.m_value.get<Buffer>());
            case List:
            case Tuple:
            case Dictionary:
            case Bag:
            case TimeSeries:
            case DataTable:
                return m_value.get<Collection>().compare(rhs.m_value.get<Collection>());
            case Exception:
                return m_value.get<Exception>().compare(rhs.m_value.get<Exception>());
            case Object:
                return m_value.get<Object>()->compare(*rhs.m_value.get<Object>());
                break;
            case Array:
                return m_value.get<Array>()->compare(*rhs.m_value.get<Array>());
                break;
            default:
                boost::throw_exception(variant_error("Unrecognised variant type " + enum_to_string(type)));
        }
    }

    boost::uint64_t variant_base::hash(enum_type_t type, boost::uint64_t seed) const
    {
        switch (type)
        {
            case None:
                return 0;
            case Any:
                return detail::hash_value(m_value.get<Any>().value(), seed);
            case String:
                return detail::hash_value(m_value.get<String>().value(), seed);
            case Int32:
                return detail::hash_value(m_value.get<Int32>(), seed);
            case UInt32:
                return detail::hash_value(m_value.get<UInt32>(), seed);
            case Int64:
                return detail::hash_value(m_value.get<Int64>(), seed);
            case UInt64:
                return detail::hash_value(m_value.get<UInt64>(), seed);
            case Float:
                return detail::hash_value(m_value.get<Float>(), seed);
            case Double:
                return detail::hash_value(m_value.get<Double>(), seed);
            case Boolean:
                return detail::hash_value(m_value.get<Boolean>(), seed);
            case Date:
                return detail::hash_value(m_value.get<Date>(), seed);
            case Time:
                return detail::hash_value(m_value.get<Time>(), seed);
            case DateTime:
                return detail::hash_value(m_value.get<DateTime>(), seed);
            case List:
            case Dictionary:
            case Bag:
            case TimeSeries:
            case DataTable:
            {
                return m_value.get<Collection>().hash(seed);
            }
            case Buffer:
            {
                return m_value.get<Buffer>()->hash(seed);
            }
            case Tuple:
            {
                return m_value.get<Tuple>().hash(seed);
            }
            case Exception:
            {
                return m_value.get<Exception>().hash(seed);
            }
            case Object:
            {
                return m_value.get<Object>()->hash(seed);
            }
            case Array:
            {
                return m_value.get<Array>()->hash(seed);
            }
            default:
            {
                boost::throw_exception(variant_error("Unrecognised variant type"));
            }
        }
    }

    void variant_base::swap(variant_base& rhs)
    {
        m_value.swap(rhs.m_value);
    }

} // namespace protean

File: bindings/cs/Protean-cs/Protean.Net/VariantBase.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Protean {

    public interface IVariantData :
        IComparable<IVariantData>
    {
        VariantBase.EnumType _Type { get; }
    }

    // None
    internal class VariantNone : IVariantData
    {
        public VariantNone()
        { }

        public VariantBase.EnumType _Type
        {
            get { return VariantBase.EnumType.None; }
        }

        public int CompareTo(IVariantData rhs)
        {
            return 0;
        }

        public override int GetHashCode()
        {
            return 0;
        }
    }

    // Any
    internal class VariantAny : IVariantData
    {
        public VariantAny() {
            Value = "";
        }

        public VariantAny(String value) {
            Value = value;
        }

        public VariantAny(VariantAny rhs) {
            Value = rhs.Value;
        }

        public VariantBase.EnumType _Type
        {
            get { return VariantBase.EnumType.Any; }
        }

        public int CompareTo(IVariantData rhs)
        {
            return Value.CompareTo(((VariantAny)rhs).Value);
        }

        public override int GetHashCode()
        {
            return Value.GetHashCode();
        }

        public string Value { get; set; }
    }

    // Buffer
    internal class VariantBuffer : IVariantData
    {
        public VariantBuffer(int capacity) {
            Value = new byte[capacity];
        }

        public VariantBuffer(byte[] value) {
            Value = value;
        }

        public VariantBuffer(VariantBuffer rhs) {
            Value = new byte[rhs.Value.Length];
            Array.Copy(rhs.Value, Value, Value.Length);
        }

        public VariantBase.EnumType _Type
        {
            get { return VariantBase.EnumType.Buffer; }
        }

        public int CompareTo(IVariantData rhs)
        {
            return SequenceComparer.Compare(Value, ((VariantBuffer)rhs).Value);
        }

        public byte[] Value { get; set; }

        public override int GetHashCode()
        {
            unchecked
            {
                var hashCode = 0x2D2816FE;
                hashCode = Value.Aggregate(hashCode, (cur, next) => cur * 397 ^ next.GetHashCode());
                return hashCode;
            }
        }
    }

    public abstract class VariantBase
    {
        [Flags]
        public enum EnumType
        {
            None                    = 0x00000001,
            Any                     = 0x00000002,
            String                  = 0x00000004,
            Boolean                 = 0x00000008,
            Int32                   = 0x00000010,
            UInt32                  = 0x00000020,
            Int64                   = 0x00000040,
            UInt64                  = 0x00000080,
            Float                   = 0x00000100,
            Double                  = 0x00000200,
            Date                    = 0x00000400,
            Time                    = 0x00000800,
            DateTime                = 0x00001000,
            List                    = 0x00002000,
            Dictionary              = 0x00004000,
            Bag                     = 0x00008000,
            Buffer                  = 0x00010000,
            Tuple                   = 0x00020000,
            Exception               = 0x00040000,
            TimeSeries              = 0x00080000,
            Object                  = 0x00100000,
            Array                   = 0x00200000,
            DataTable               = 0x00400000,

            // The following enums represent groupings of variants
            Integer                 = Boolean | Int32 | UInt32 | Int64 | UInt64,
            Number                  = Integer | Float | Double,
            Primitive               = Number | Date | Time | DateTime | Any | String,
            Sequence                = List | Tuple,
            Mapping                 = Dictionary | Bag,
            Collection              = Sequence | Mapping | TimeSeries
        };

        internal IVariantData Value { get; set; }

        protected VariantBase()
        {
            Value = new VariantNone();
        }

        protected VariantBase(EnumType type, string value)
        {
            switch (type)
            {
                case EnumType.Any:
                    Value = new VariantAny(value);
                    break;
                case EnumType.String:
                    Value = new VariantPrimitive<String>(value);
                    break;
                case EnumType.Boolean:
                    Value = new VariantPrimitive<bool>(ParseBoolean(value));
                    break;
                case EnumType.Int32:
                    Value = new VariantPrimitive<Int32>(Int32.Parse(value));
                    break;
                case EnumType.UInt32:
                    Value = new VariantPrimitive<UInt32>(UInt32.Parse(value));
                    break;
                case EnumType.Int64:
                    Value = new VariantPrimitive<Int64>(Int64.Parse(value));
                    break;
                case EnumType.UInt64:
                    Value = new VariantPrimitive<UInt64>(UInt64.Parse(value));
                    break;
                case EnumType.Float:
                    Value = new VariantPrimitive<float>(ParseSingle(value));
                    break;
                case EnumType.Double:
                    Value = new VariantPrimitive<double>(ParseDouble(value));
                    break;
                case EnumType.DateTime:
                    Value = new VariantPrimitive<DateTime>(ParseDateTime(value));
                    break;
                case EnumType.Time:
                    Value = new VariantPrimitive<TimeSpan>(ParseTime(value));
                    break;
                default:
                    throw new VariantException(string.Format("Unable to construct variant of type {0} from string", Type));
            }
        }

        protected VariantBase(EnumType type, int size)
        {
            switch (type)
            {
            case EnumType.None:
                Value = new VariantNone();
                break;
            case EnumType.Any:
                Value = new VariantAny();
                break;
            case EnumType.Boolean:
                Value = new VariantPrimitive<bool>(false);
                break;
            case EnumType.Int32:
                Value = new VariantPrimitive<Int32>(0);
                break;
            case EnumType.UInt32:
                Value = new VariantPrimitive<UInt32>(0);
                break;
            case EnumType.Int64:
                Value = new VariantPrimitive<Int64>(0);
                break;
            case EnumType.UInt64:
                Value = new VariantPrimitive<UInt64>(0);
                break;
            case EnumType.String:
                Value = new VariantPrimitive<String>("");
                break;
            case EnumType.DateTime:
                Value = new VariantPrimitive<DateTime>(new DateTime());
                break;
            case EnumType.Time:
                Value = new VariantPrimitive<TimeSpan>(new TimeSpan());
                break;
            case EnumType.List:
                Value = new VariantList(size);
                break;
            case EnumType.TimeSeries:
                Value = new VariantTimeSeries(size);
                break;
            case EnumType.Dictionary:
                Value = new VariantDictionary();
                break;
            case EnumType.Bag:
                Value = new VariantBag();
                break;
            case EnumType.Tuple:
                Value = new VariantTuple(size);
                break;
            case EnumType.Buffer:
                Value = new VariantBuffer(size);
                break;
            default:
                throw new VariantException("Cannot default construct variant of type: " + type.ToString());
            }
        }

        protected VariantBase(VariantBase rhs)
        {
            EnumType type = rhs.Type;
            switch (type)
            {
                case EnumType.None:
                    Value = new VariantNone();
                    break;
                case EnumType.Boolean:
                    Value = new VariantPrimitive<bool>(rhs.Value as VariantPrimitive<bool>);
                    break;
                case EnumType.Double:
                    Value = new VariantPrimitive<double>(rhs.Value as VariantPrimitive<double>);
                    break;
                case EnumType.Int32:
                    Value = new VariantPrimitive<Int32>(rhs.Value as VariantPrimitive<Int32>);
                    break;
                case EnumType.UInt32:
                    Value = new VariantPrimitive<UInt32>(rhs.Value as VariantPrimitive<UInt32>);
                    break;
                case EnumType.Int64:
                    Value = new VariantPrimitive<Int64>(rhs.Value as VariantPrimitive<Int64>);
                    break;
                case EnumType.UInt64:
                    Value = new VariantPrimitive<UInt64>(rhs.Value as VariantPrimitive<UInt64>);
                    break;
                case EnumType.DateTime:
                    Value = new VariantPrimitive<DateTime>(rhs.Value as VariantPrimitive<DateTime>);
                    break;
                case EnumType.Time:
                    Value = new VariantPrimitive<TimeSpan>(rhs.Value as VariantPrimitive<TimeSpan>);
                    break;
                case EnumType.String:
                    Value = new VariantPrimitive<String>(rhs.Value as VariantPrimitive<String>);
                    break;
                case EnumType.Any:
                    Value = new VariantAny(rhs.Value as VariantAny);
                    break;
                case EnumType.List:
                    Value = new VariantList(rhs.Value as VariantList);
                    break;
                case EnumType.Dictionary:
                    Value = new VariantDictionary(rhs.Value as VariantDictionary);
                    break;
                case EnumType.Bag:
                    Value = new VariantBag(rhs.Value as VariantBag);
                    break;
                case EnumType.TimeSeries:
                    Value = new VariantTimeSeries(rhs.Value as VariantTimeSeries);
                    break;
                case EnumType.Tuple:
                    Value = new VariantTuple(rhs.Value as VariantTuple);
                    break;
                case EnumType.Buffer:
                    Value = new VariantBuffer(rhs.Value as VariantBuffer);
                    break;
                case EnumType.Exception:
                    Value = new VariantExceptionInfo(rhs.Value as VariantExceptionInfo);
                    break;
                default:
                    throw new VariantException("Cannot copy variant of type: " + type.ToString());
            }
        }

        public EnumType Type
        {
            get
            {
                return Value._Type;
            }
        }

        // Is/As only work for primitive types
        public T As<T>() where T : IComparable<T>
        {
            if (Value is VariantPrimitive<T>)
            {
                return ((VariantPrimitive<T>)Value).Value;
            } 
            else if (Value is VariantAny)
            {
                string str = ((VariantAny)Value).Value;
                return VariantPrimitiveBase.Parse<T>(str);
            }
            else
            {
                throw new VariantException("Attempt to fetch " + typeof(T).Name + " from " + Type.ToString() + " variant.");
            }
        }

        public bool Is<T>() where T : IComparable<T>
        {
            return (Value is VariantPrimitive<T>);
        }

        public bool Is(EnumType type)
        {
            return (type & Value._Type)!=0;
        }

        public static float ParseSingle(string value)
        {
            float singleValue;

            try
            {
                singleValue = float.Parse(value);
            }
            catch (FormatException)
            {
                if (!s_nonRoundTripSingleSpecialValues.TryGetValue(value, out singleValue))
                {
                    throw;
                }
            }
            catch (OverflowException)
            {
                if (!s_nonRoundTripSingleSpecialValues.TryGetValue(value, out singleValue))
                {
                    throw;
                }
            }

            return singleValue;
        }

        public static double ParseDouble(string value)
        {
            double doubleValue;

            try
            {
                doubleValue = double.Parse(value);
            }
            catch (FormatException)
            {
                if (!s_nonRoundTripDoubleSpecialValues.TryGetValue(value, out doubleValue))
                {
                    throw;
                }
            }
            catch (OverflowException)
            {
                if (!s_nonRoundTripDoubleSpecialValues.TryGetValue(value, out doubleValue))
                {
                    throw;
                }
            }

            return doubleValue;
        }

        public static bool ParseBoolean(string str)
        {
            switch (str)
            {
                case "true":
                case "1":
                    return true;
                case "false":
                case "0":
                    return false;
                default:
                    throw new VariantException("Illegal format for Boolean, expecting true/1 or false/0, encountered " + str);
            }
        }

        public static DateTime ParseDateTime(string str)
        {
            return DateTime.Parse(str);
        }

        public static TimeSpan ParseTime(string str)
        {
            return TimeSpan.Parse(str);
        }

        public static string ToString(bool arg)
        {
            return arg ? "true" : "false";
        }


        public static string ToString(float arg)
        {
            if (float.IsPositiveInfinity(arg))
            {
                return "INF";
            }
            else if(float.IsNegativeInfinity(arg))
            {
                return "-INF";
            }
            else
            {
                return arg.ToString("R");
            }
        }


        public static string ToString(double arg)
        {
            if (double.IsPositiveInfinity(arg))
            {
                return "INF";
            }
            else if (double.IsNegativeInfinity(arg))
            {
                return "-INF";
            }
            else
            {
                return arg.ToString("R");
            }
        }


        public static string ToString(int arg)
        {
            if (int.MaxValue == arg)
            {
                return "INF";
            }
            else if (int.MinValue == arg)
            {
                return "-INF";
            }
            else
            {
                return arg.ToString();
            }
        }


        public static string ToString(uint arg)
        {
            if (uint.MaxValue == arg)
            {
                return "INF";
            }
            else if (uint.MinValue == arg)
            {
                return "-INF";
            }
            else
            {
                return arg.ToString();
            }
        }


        public static string ToString(ulong arg)
        {
            if (ulong.MaxValue == arg)
            {
                return "INF";
            }
            else if (uint.MinValue == arg)
            {
                return "-INF";
            }
            else
            {
                return arg.ToString();
            }
        }


        public static string ToString(long arg)
        {
            if (long.MaxValue == arg)
            {
                return "INF";
            }
            else if (long.MinValue == arg)
            {
                return "-INF";
            }
            else
            {
                return arg.ToString();
            }
        }
        

        public static string ToString(DateTime arg)
        {
            if (arg.Millisecond == 0)
            {
                return arg.ToString("yyyy-MM-ddTHH:mm:ss");
            }
            else
            {
                return arg.ToString("yyyy-MM-ddTHH:mm:ss.fff");
            }
        }

        public static string ToString(TimeSpan arg)
        {
            return $"{(arg < TimeSpan.Zero ? "-" : "")}{arg:hh\\:mm\\:ss}{(arg.Milliseconds != 0 ? arg.ToString(@"\.FFF") : "")}";
        }

        public override int GetHashCode()
        {
            unchecked
            {
                return Type.GetHashCode() * 397 ^ Value.GetHashCode();
            }
        }

        private readonly static Dictionary<string, float> s_nonRoundTripSingleSpecialValues =
            new Dictionary<string, float>
                {
                    {float.MaxValue.ToString(), float.MaxValue},
                    {float.MinValue.ToString(), float.MinValue},
                    {float.PositiveInfinity.ToString(), float.PositiveInfinity},
                    {float.NegativeInfinity.ToString(), float.NegativeInfinity},
                    {"INF", float.PositiveInfinity},
                    {"-INF", float.NegativeInfinity},
                };

        private readonly static Dictionary<string, double> s_nonRoundTripDoubleSpecialValues =
            new Dictionary<string, double>
                {
                    {double.MaxValue.ToString(), double.MaxValue},
                    {double.MinValue.ToString(), double.MinValue},
                    {double.PositiveInfinity.ToString(), double.PositiveInfinity},
                    {double.NegativeInfinity.ToString(), double.NegativeInfinity},
                    {"INF", double.PositiveInfinity},
                    {"-INF", double.NegativeInfinity},
                };
    }

} // Protean
